core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+                $scope, $uibModalInstance, AlertsService,
+                LearnerActionRenderService,
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+                MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+              $scope.playthroughIndex = playthroughIndex;
+
+              // This is the index of the first action in the range of actions
+              // displayed. When this index reaches 0, we've displayed all the
+              // learner actions.
+              let indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return indexOfFirstDisplayedAction !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction ===
+                    playthrough.actions.length) {
+                  previousStateName = (
+                    playthrough.actions[
+                      0].actionCustomizationArgs.state_name.value);
+                } else {
+                  let action = playthrough.actions[indexOfFirstDisplayedAction];
+                  previousStateName =
+                    action.actionCustomizationArgs.state_name.value;
+                }
+                for (i = indexOfFirstDisplayedAction; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+
+                  if (currentStateName !== previousStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible.
+                if (i === -1) {
```

Comment:
It seems to me that this shouldn't be a special case. Is there an off-by-one error in this method somewhere? I note that there are "indexOfFirstDisplayedAction + 1" items being iterated over in the for loop, which seems like 1 more than the number I would expect.

Reply:
@seanlip This case exists because sometimes there might be the case when we iterate over the final display block (the block that will be displayed last), and the number of actions in that block are less than the threshold constant. In this case, the current action index does not get updated to 0 and we exit the loop. So, to handle this, we keep that check outside that loop for when all the actions have been iterated over.

Reply:
Ah, got it, thanks. I suggest updating the comment to explain the situation in which this occurs.

Still ... shouldn't the for loop start at indexOfFirstDisplayedAction - 1?

Reply:
@seanlip Yep the loop could start at index - 1, but anyways the if condition would not be triggered for the first iteration, so its pretty much the same thing. So we can start it at Index - 1 too.

Reply:
OK, let's do that. It's conceptually clearer, and that's important.

Reply:
Done.